File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.cc
#include <iostream>
#include <stack>
#include <string>

using namespace std;

int main() {
	ios::sync_with_stdio(0);
	cin.tie(0);
	while (true) {
		string input;
		getline(cin, input);
		if (input == ".") break;
		stack<char> s;
		bool isValid = true;
		for(auto c : input) {
			if (c == '(' || c == '[') s.push(c);
			else if (c == ')') {
				if (s.empty() || s.top() != '(') {
					isValid = false;
					break;
				}
				s.pop();
			}
			else if (c == ']') {
				if (s.empty() || s.top() != '[') {
					isValid = false;
					break;
				}
				s.pop();
			}
		}
		if (!s.empty()) isValid = false;
		if (isValid) cout << "yes\n";
		else cout << "no\n";
	}
}

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.py
import sys

input = sys.stdin.readline

while True:
    line = input().rstrip()
    if line == ".":
        break
    stack = []
    is_valid = True  
    for c in line:
        if c == '(' or c == '[':
            stack.append(c)
        elif c == ')':
            if not stack or stack[-1] != '(':
                is_valid = False
                break
            stack.pop()
        elif c == ']':
            if not stack or stack[-1] != '[':
                is_valid = False
                break
            stack.pop()
    if stack:
        is_valid = False
    print("yes" if is_valid else "no")
